File: sirius-core/src/main/java/org/starcoin/sirius/core/AugmentedMerkleTree.java
package org.starcoin.sirius.core;

import org.apache.commons.lang3.RandomUtils;
import org.starcoin.core.*;
import org.starcoin.core.MerklePath.Direction;
import org.starcoin.sirius.core.AugmentedMerklePath.AugmentedMerklePathNode;
import org.starcoin.proto.Starcoin.ProtoAugmentedMerkleTreeNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class AugmentedMerkleTree {

    public static final class AugmentedMerkleTreeNode
            implements Hashable, ProtobufCodec<ProtoAugmentedMerkleTreeNode> {

        private transient AugmentedMerkleTreeNode parent;
        private transient AugmentedMerkleTreeNode left;
        private transient AugmentedMerkleTreeNode right;
        private transient Hash hash;

        private long offset;
        private NodeInformation information;
        private AccountInformation account;
        private long allotment;

        public AugmentedMerkleTreeNode() {
            this(0, NodeInformation.EMPTY_NODE, 0);
        }

        public AugmentedMerkleTreeNode(long offset, AccountInformation account, long allotment) {
            this.offset = offset;
            this.account = account;
            this.allotment = allotment;
        }

        public AugmentedMerkleTreeNode(long offset, NodeInformation node, long allotment) {
            this.offset = offset;
            this.information = node;
            this.allotment = allotment;
        }

        public AugmentedMerkleTreeNode(AugmentedMerkleTreeNode prev, AccountInformation account) {
            this.offset = prev == null ? 0 : prev.offset + prev.allotment;
            this.account = account;
            this.allotment = account.getAllotment();
        }

        public AugmentedMerkleTreeNode(AugmentedMerkleTreeNode left) {
            this(left, (AugmentedMerkleTreeNode) null);
        }

        public AugmentedMerkleTreeNode(ProtoAugmentedMerkleTreeNode proto) {
            this.unmarshalProto(proto);
        }

        public AugmentedMerkleTreeNode(AugmentedMerkleTreeNode left, AugmentedMerkleTreeNode right) {
            if (right == null) {
                right =
                        new AugmentedMerkleTreeNode(
                                left.offset + left.allotment, AccountInformation.EMPTY_ACCOUNT, 0);
            }
            this.left = left;
            this.right = right;
            this.left.parent = this;
            this.right.parent = this;
            this.offset = left.offset;
            this.allotment = left.allotment + right.allotment;
            this.information = new NodeInformation(left.hash(), right.offset, right.hash());
        }

        @Override
        public Hash hash() {
            if (this.hash != null) {
                return this.hash;
            }
            this.hash = Hash.of(this.toProto().toByteArray());
            return this.hash;
        }

        public boolean isLeafNode() {
            return this.left == null && this.right == null && this.account != null;
        }

        public boolean isNnternalNode() {
            return this.information != null;
        }

        public AccountInformation getAccount() {
            return this.account;
        }

        public NodeInformation getInformation() {
            return information;
        }

        public long getOffset() {
            return offset;
        }

        public long getAllotment() {
            return allotment;
        }

        public AugmentedMerkleTreeNode getParent() {
            return parent;
        }

        public AugmentedMerkleTreeNode getLeft() {
            return left;
        }

        public AugmentedMerkleTreeNode getRight() {
            return right;
        }

        public AugmentedMerkleTreeNode getSibling() {
            if (this.parent == null) {
                return null;
            }
            return this == this.parent.left ? this.parent.right : this.parent.left;
        }

        public Direction getDirection() {
            if (this.parent == null) {
                return null;
            }
            return this == this.parent.left ? MerklePath.Direction.LEFT : MerklePath.Direction.RIGHT;
        }

        @Override
        public ProtoAugmentedMerkleTreeNode marshalProto() {
            ProtoAugmentedMerkleTreeNode.Builder builder =
                    ProtoAugmentedMerkleTreeNode.newBuilder().setOffset(this.offset);
            if (this.account != null) {
                builder.setAccount(this.account.toProto());
            }
            if (this.information != null) {
                builder.setNode(this.information.toProto());
            }
            return builder.setAllotment(this.allotment).build();
        }

        @Override
        public void unmarshalProto(ProtoAugmentedMerkleTreeNode proto) {
            this.offset = proto.getOffset();
            this.account =
                    proto.hasAccount()
                            ? AccountInformation.generateAccountInformation(proto.getAccount())
                            : null;
            this.information =
                    proto.hasNode() ? NodeInformation.generateNodeInformation(proto.getNode()) : null;
            this.allotment = proto.getAllotment();
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) {
                return true;
            }
            if (!(o instanceof AugmentedMerkleTreeNode)) {
                return false;
            }
            AugmentedMerkleTreeNode treeNode = (AugmentedMerkleTreeNode) o;
            return offset == treeNode.offset
                    && allotment == treeNode.allotment
                    && Objects.equals(information, treeNode.information)
                    && Objects.equals(account, treeNode.account);
        }

        @Override
        public int hashCode() {
            return Objects.hash(offset, information, account, allotment);
        }

        public static AugmentedMerkleTreeNode generateTreeNode(ProtoAugmentedMerkleTreeNode proto) {
            AugmentedMerkleTreeNode treeNode = new AugmentedMerkleTreeNode();
            treeNode.unmarshalProto(proto);
            return treeNode;
        }

        public String toString() {
            return this.toJson();
        }
    }

    private int eon = 1;
    private AugmentedMerkleTreeNode root;

    public AugmentedMerkleTree(int eon) {
        this.eon = eon;
        this.root = new AugmentedMerkleTreeNode(0, AccountInformation.EMPTY_ACCOUNT, 0);
    }

    public AugmentedMerkleTree(int eon, List<AccountInformation> accountInformations) {
        this.eon = eon;
        this.root = buildRoot(buildTreeNodes(accountInformations));
    }

    public int getEon() {
        return eon;
    }

    public AugmentedMerkleTreeNode getRoot() {
        return root;
    }

    public Hash hash() {
        return this.root.hash();
    }

    public long getOffset() {
        return this.root.offset;
    }

    public long getAllotment() {
        return this.root.allotment;
    }

    public NodeInformation getInformation() {
        return this.root.information;
    }

    public Optional<AugmentedMerkleTreeNode> findTreeNode(Hash nodeHash) {
        return findTreeNode(this.root, node -> node.hash().equals(nodeHash));
    }

    public Optional<AccountInformation> findAccountInfomation(BlockAddress blockAddress) {
        Optional<AugmentedMerkleTreeNode> treeNode = this.findLeafNode(blockAddress);
        if (treeNode.isPresent()) {
            return Optional.of(treeNode.get().getAccount());
        } else {
            return Optional.empty();
        }
    }

    public Optional<AugmentedMerkleTreeNode> findLeafNode(BlockAddress blockAddress) {
        return this.findLeafNode(Hash.of(blockAddress.toBytes()));
    }

    private Optional<AugmentedMerkleTreeNode> findLeafNode(Hash blockAddressHash) {
        return findTreeNode(
                this.root,
                node -> node.isLeafNode() && node.getAccount().getAddress().equals(blockAddressHash));
    }

    private Optional<AugmentedMerkleTreeNode> findTreeNode(
            AugmentedMerkleTreeNode node, Predicate<AugmentedMerkleTreeNode> predicate) {
        if (node == null) {
            return Optional.empty();
        }
        if (predicate.test(node)) {
            return Optional.of(node);
        }
        Optional<AugmentedMerkleTreeNode> found = findTreeNode(node.left, predicate);
        if (found.isPresent()) {
            return found;
        } else {
            return findTreeNode(node.right, predicate);
        }
    }

    // just for test.
    public AugmentedMerklePath getRandommProof() {
        return this.getMembershipProof(this.randomLeafNode().getAccount().getAddress());
    }

    public AugmentedMerklePath getMembershipProof(Hash blockAddressHash) {
        Optional<AugmentedMerkleTreeNode> nodeOptional = this.findLeafNode(blockAddressHash);
        if (!nodeOptional.isPresent()) {
            return null;
        }
        AugmentedMerkleTreeNode node = nodeOptional.get();
        AugmentedMerklePath path = new AugmentedMerklePath(this.eon);
        path.append(node, node.getDirection());
        AugmentedMerkleTreeNode siblingNode = node.getSibling();
        path.append(siblingNode, siblingNode.getDirection());

        AugmentedMerkleTreeNode parent = node.getParent();
        while (parent.getParent() != null) {
            siblingNode = parent.getSibling();
            path.append(siblingNode, siblingNode.getDirection());
            parent = parent.getParent();
        }
        return path;
    }

    public AugmentedMerklePath getMembershipProof(BlockAddress blockAddress) {
        return this.getMembershipProof(Hash.of(blockAddress.toBytes()));
    }

    public AugmentedMerkleTreeNode randomLeafNode() {
        AugmentedMerkleTreeNode node = randomChild(this.root);
        while (!node.isLeafNode()) {
            node = randomChild(node);
        }
        return node;
    }

    private AugmentedMerkleTreeNode randomChild(AugmentedMerkleTreeNode node) {
        if (node.isLeafNode()) {
            return node;
        }
        if (RandomUtils.nextBoolean()) {
            return node.left;
        } else {
            return node.right;
        }
    }

    private static List<AugmentedMerkleTreeNode> buildTreeNodes(
            List<AccountInformation> accountInformationList) {
        final AugmentedMerkleTreeNode[] prev = {null};
        return accountInformationList
                .stream()
                .map(
                        accountInformation -> {
                            AugmentedMerkleTreeNode node =
                                    new AugmentedMerkleTreeNode(prev[0], accountInformation);
                            prev[0] = node;
                            return node;
                        })
                .collect(Collectors.toList());
    }

    private static AugmentedMerkleTreeNode buildRoot(List<AugmentedMerkleTreeNode> leaves) {
        if (leaves.isEmpty()) {
            return new AugmentedMerkleTreeNode();
        }
        List<AugmentedMerkleTreeNode> mergedLeaves = new ArrayList<>();
        for (int i = 0, n = leaves.size(); i < n; i++) {
            if (i < n - 1) {
                mergedLeaves.add(new AugmentedMerkleTreeNode(leaves.get(i), leaves.get(i + 1)));
                i++;
            } else {
                mergedLeaves.add(new AugmentedMerkleTreeNode(leaves.get(i)));
            }
        }

        if (mergedLeaves.size() > 1) {
            return buildRoot(mergedLeaves);
        } else {
            return mergedLeaves.get(0);
        }
    }

    public static AugmentedMerkleTreeNode buildRoot(AugmentedMerklePath path) {
        List<AugmentedMerklePathNode> nodes = path.getNodes();

        AugmentedMerkleTreeNode node = nodes.get(0).getNode();

        for (int i = 1; i < nodes.size(); i++) {
            AugmentedMerklePathNode pathNode = nodes.get(i);
            if (pathNode.getDirection() == MerklePath.Direction.LEFT) {
                node = new AugmentedMerkleTreeNode(pathNode.getNode(), node);
            } else {
                node = new AugmentedMerkleTreeNode(node, pathNode.getNode());
            }
        }
        return node;
    }

    public static boolean verifyMembershipProof(
            AugmentedMerkleTreeNode root, AugmentedMerklePath path) {
        AugmentedMerkleTreeNode rootBuild = buildRoot(path);
        return rootBuild.hash().equals(root.hash());
    }

    public static AugmentedMerkleTree random(int count) {
        int eon = 1;
        List<AccountInformation> accountInformationList = new ArrayList<>();

        for (int i = 0; i < count; i++) {
            int allotment = RandomUtils.nextInt(0, 10000);
            int receive = RandomUtils.nextInt(0, 10000);
            int send = RandomUtils.nextInt(0, allotment + receive);
            AccountInformation a =
                    new AccountInformation(
                            BlockAddress.random(), allotment, new Update(eon, 0, send, receive, Hash.random()));
            accountInformationList.add(a);
        }

        AugmentedMerkleTree tree = new AugmentedMerkleTree(eon, accountInformationList);
        return tree;
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/AugmentedMerkleTree.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import org.apache.commons.lang3.RandomUtils
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.serialization.ProtobufSchema
import java.util.*
import java.util.stream.Collectors

sealed class AMTNodeInfo : SiriusObject()

@ProtobufSchema(Starcoin.AMTInternalNodeInfo::class)
@Serializable
data class AMTInternalNodeInfo(
    @SerialId(1) val left: Hash,
    @SerialId(2) val offset: Long,
    @SerialId(3) val right: Hash
) : AMTNodeInfo() {
    constructor(left: Hashable, offset: Long, right: Hashable) : this(left.hash(), offset, right.hash())

    companion object :
        SiriusObjectCompanion<AMTInternalNodeInfo, Starcoin.AMTInternalNodeInfo>(AMTInternalNodeInfo::class) {
        val DUMMY_NODE = AMTInternalNodeInfo(Hash.EMPTY_DADA_HASH, 0, Hash.EMPTY_DADA_HASH)

        override fun mock(): AMTInternalNodeInfo {
            return AMTInternalNodeInfo(Hash.random(), RandomUtils.nextLong(), Hash.random())
        }
    }
}

@ProtobufSchema(Starcoin.AMTLeafNodeInfo::class)
@Serializable
data class AMTLeafNodeInfo(
    @SerialId(1) val addressHash: Hash,
    @SerialId(2) val update: Update
) : AMTNodeInfo() {
    companion object : SiriusObjectCompanion<AMTLeafNodeInfo, Starcoin.AMTLeafNodeInfo>(AMTLeafNodeInfo::class) {
        val DUMMY_NODE = AMTLeafNodeInfo(Hash.EMPTY_DADA_HASH, Update.DUMMY_UPDATE)
        override fun mock(): AMTLeafNodeInfo {
            return AMTLeafNodeInfo(Hash.random(), Update.mock())
        }
    }
}

@ProtobufSchema(Starcoin.AMTProof::class)
@Serializable
data class AMTProof(@SerialId(1) val path: AMTPath, @SerialId(2) val leaf: AMTPathLeafNode) : SiriusObject() {
    companion object : SiriusObjectCompanion<AMTProof, Starcoin.AMTProof>(AMTProof::class) {
        override fun mock(): AMTProof {
            return AMTProof(AMTPath.mock(), AMTPathLeafNode.mock())
        }
    }
}

class AugmentedMerkleTree(
    val eon: Int,
    val root: AMTNode = AMTNode()
) {


    val offset: Long
        get() = this.root.offset

    val allotment: Long
        get() = this.root.allotment

    val info: AMTNodeInfo?
        get() = this.root.info

    // just for test.
    val randommProof: AMTProof?
        get() = this.getMembershipProof((this.randomLeafNode()?.info as AMTLeafNodeInfo?)?.addressHash)


    constructor(eon: Int, accounts: List<HubAccount>) : this(
        eon,
        buildRoot(buildTreeNodes(accounts))
    )

    fun hash(): Hash {
        return this.root.hash()
    }

    fun findTreeNode(nodeHash: Hash): AMTNode? {
        return findTreeNode(this.root) { node -> node.hash() == nodeHash }
    }

    fun findLeafNode(address: Address?): AMTNode? {
        return this.findLeafNode(address?.hash())
    }

    fun findLeafNode(addressHash: Hash?): AMTNode? {
        return findTreeNode(
            this.root
        ) { node -> node.isLeafNode && (node.info as AMTLeafNodeInfo).addressHash == addressHash }
    }

    private fun findTreeNode(
        node: AMTNode?, predicate: (AMTNode) -> Boolean
    ): AMTNode? {
        return when {
            node == null -> null
            predicate(node) -> node
            else -> findTreeNode(node.left, predicate) ?: findTreeNode(node.right, predicate)

        }
    }

    fun getMembershipProof(addressHash: Hash?): AMTProof? {
        val leaf = this.findLeafNode(addressHash) ?: return null

        var siblingNode = leaf.sibling ?: return null

        val path = AMTPath(
            this.eon,
            siblingNode.toAMTPathNode() as AMTPathLeafNode
        )
        val proof = AMTProof(path, leaf.toAMTPathNode() as AMTPathLeafNode)

        var parent = leaf.parent ?: return proof
        while (parent.parent != null) {
            siblingNode = parent.sibling ?: return proof
            path.append(siblingNode)
            parent = parent.parent ?: return proof
        }
        return proof
    }

    fun getMembershipProof(address: Address?): AMTProof? {
        return this.getMembershipProof(address?.hash())
    }

    fun randomLeafNode(): AMTNode? {
        var node = randomChild(this.root) ?: return null
        while (!node.isLeafNode) {
            node = randomChild(node) ?: return null
        }
        return node
    }

    private fun randomChild(node: AMTNode): AMTNode? {
        return when {
            node.isLeafNode -> node
            RandomUtils.nextBoolean() -> node.left
            else -> node.right
        }
    }

    companion object {

        private fun buildTreeNodes(
            accounts: List<HubAccount>
        ): List<AMTNode> {
            val prev = arrayOfNulls<AMTNode>(1)
            return accounts
                .stream()
                .map { account ->
                    val node = AMTNode(
                        prev[0],
                        AMTLeafNodeInfo(account.address.hash(), account.update),
                        account.allotment
                    )
                    prev[0] = node
                    node
                }
                .collect(Collectors.toList())
        }

        private fun buildRoot(leaves: List<AMTNode>): AMTNode {
            if (leaves.isEmpty()) {
                return AMTNode()
            }
            val mergedLeaves = ArrayList<AMTNode>()
            var i = 0
            val n = leaves.size
            while (i < n) {
                if (i < n - 1) {
                    mergedLeaves.add(AMTNode(leaves[i], leaves[i + 1]))
                    i++
                } else {
                    mergedLeaves.add(AMTNode(leaves[i]))
                }
                i++
            }

            return if (mergedLeaves.size > 1) {
                buildRoot(mergedLeaves)
            } else {
                mergedLeaves[0]
            }
        }

        fun buildRoot(path: AMTPath, leafNode: AMTPathLeafNode): AMTNode {

            var node = if (path.leaf.direction == MerklePath.Direction.LEFT) AMTNode(
                AMTNode(path.leaf), AMTNode(leafNode)
            ) else AMTNode(AMTNode(leafNode), AMTNode(path.leaf))

            for (i in 0 until path.size) {
                val pathNode = path[i]
                node = when {
                    pathNode.direction == MerklePath.Direction.LEFT -> AMTNode(
                        AMTNode(
                            pathNode
                        ), node
                    )
                    else -> AMTNode(
                        node,
                        AMTNode(pathNode)
                    )
                }
            }
            return node
        }

        fun verifyMembershipProof(root: AMTPathInternalNode?, proof: AMTProof?): Boolean {
            return verifyMembershipProof(root, proof?.path, proof?.leaf)
        }

        fun verifyMembershipProof(
            root: AMTNode?, proof: AMTProof?
        ): Boolean {
            return this.verifyMembershipProof(root?.toAMTPathNode() as AMTPathInternalNode?, proof?.path, proof?.leaf)
        }

        fun verifyMembershipProof(
            root: AMTNode?, path: AMTPath?, leaf: AMTNode?
        ): Boolean {
            return this.verifyMembershipProof(
                root?.toAMTPathNode() as AMTPathInternalNode?,
                path,
                leaf?.toAMTPathNode() as AMTPathLeafNode
            )
        }

        fun verifyMembershipProof(root: AMTPathInternalNode?, path: AMTPath?, leaf: AMTPathLeafNode?): Boolean {
            return when {
                root == null || path == null || leaf == null -> false
                else -> {
                    val buildRoot = buildRoot(path, leaf)
                    return buildRoot.hash() == root.hash() && buildRoot.offset == root.offset && buildRoot.allotment == root.allotment
                }
            }
        }

        fun random(): AugmentedMerkleTree {
            return random(RandomUtils.nextInt(1, 100))
        }

        fun random(count: Int): AugmentedMerkleTree {
            val eon = 1
            val accounts = mutableListOf<HubAccount>()

            for (i in 0 until count) {
                accounts.add(HubAccount.mock())
            }

            return AugmentedMerkleTree(eon, accounts)
        }
    }
}

class AMTNode(
    val offset: Long = 0,
    val info: AMTNodeInfo = AMTLeafNodeInfo.DUMMY_NODE,
    val allotment: Long = 0
) : CachedHash() {

    var parent: AMTNode? = null
        private set

    var left: AMTNode? = null
        private set

    var right: AMTNode? = null
        private set


    constructor(
        prev: AMTNode?,
        info: AMTLeafNodeInfo,
        allotment: Long
    ) : this(if (prev == null) 0 else prev.offset + prev.allotment, info, allotment)

    constructor(
        left: AMTNode, right: AMTNode = AMTNode(
            left.offset + left.allotment, AMTLeafNodeInfo.DUMMY_NODE, 0
        )
    ) : this(left.offset, AMTInternalNodeInfo(left, right.offset, right), left.allotment + right.allotment) {
        left.parent = this
        right.parent = this
        this.left = left
        this.right = right
    }

    constructor(pathNode: AMTPathLeafNode) : this(
        pathNode.offset,
        pathNode.nodeInfo,
        pathNode.allotment
    )

    constructor(pathNode: AMTPathInternalNode) : this(
        pathNode.offset,
        pathNode.nodeInfo,
        pathNode.allotment
    )

    val isLeafNode: Boolean
        get() = this.left == null && this.right == null && this.info is AMTLeafNodeInfo

    val isInternalNode: Boolean
        get() = this.info is AMTInternalNodeInfo

    val sibling: AMTNode?
        get() {
            return when {
                this.parent == null -> null
                this === this.parent?.left -> this.parent?.right
                else -> this.parent?.left
            }
        }

    val direction: MerklePath.Direction
        get() {
            return when {
                this.parent == null -> MerklePath.Direction.ROOT
                this === this.parent?.left -> MerklePath.Direction.LEFT
                else -> MerklePath.Direction.RIGHT
            }
        }

    override fun doHash(): Hash {
        return Hash.of(this.toAMTPathNode())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AMTNode) return false

        if (offset != other.offset) return false
        if (info != other.info) return false
        if (allotment != other.allotment) return false

        return true
    }

    override fun hashCode(): Int {
        var result = offset.hashCode()
        result = 31 * result + info.hashCode()
        result = 31 * result + allotment.hashCode()
        return result
    }

    fun toAMTPathNode(): AMTPathNode {
        return when (this.info) {
            is AMTInternalNodeInfo -> AMTPathInternalNode(
                this.info,
                this.direction,
                this.offset,
                this.allotment
            )
            is AMTLeafNodeInfo -> AMTPathLeafNode(
                this.info,
                this.direction,
                this.offset,
                this.allotment
            )
        }
    }

}

